### DISCUSSION TITLE
Welcome to spoota-internship-roadmap Discussions!

### DISCUSSION THREAD

**Discussion by VahidFarahmandian** [VahidFarahmandian/spoota-internship-roadmap#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by CesSolutions** [comment]
Hello dear @VahidFarahmandian .
I am trainee in C# and ASP.Net core. Indeed I want to change my job as programmer. I made many small applications for myself but, nowadays I am trying to put some of my application on **GitHub** and do more projects as resume to present my ability. I hope to be a junior programmer very soon. thanks

[This is my public repositories: ](https://github.com/CesSolutions)

**Reply by VahidFarahmandian** [reply]
Dear @CesSolutions 
I will surely check your repo and write my feedback about it.

**Reply by VahidFarahmandian** [reply]
@CesSolutions I have checked your codes. They are awesome, but here are some notes:

1. In ReadMe.md: Your images are broken. it is better to check them.
2. You have done great job, but your methods have very long lines of code. you need to redecorate your methods to have short-length methods so that your codes readability will be increased. (for example: `CesControlBase/ArrangeControls`)
3. In `GlobalFunctions` you have unused commented codes. you can remove them as you are using version control.
4. You need to write your comments in Farsi or in English. do not mix them up.
5. In `GlobalFunctions` method: `ConvertToTypedList` has do almost nothing.
6. Class `StringExtnsions` is misspelled
7. I recommend not to write multiple classes in one file. You can separate them to increase the readability of codes.
8. It would be better if you provide some unit tests for your codes
9. You can put all your `png` files in a folder called `asset` to clarify your files structure

**Reply by CesSolutions** [reply]
Hello dear  @VahidFarahmandian

I divided methods which consisted of a lot of code lines to separate smaller methods. This refactoring made debugging easier. So, more dividing will done while developing this project. Thanks

**Reply by VahidFarahmandian** [reply]
Great job. @MojganHashemi @HosseinDinarvand @mahshadch1381 
I have mentioned our team members to check further your code.

**Comment by CesSolutions** [comment]
Dear @VahidFarahmandian 
I am very grateful that my repo was seen by a senior.

1. I checked ReadMe.md with different browser and mobile but all images loaded completely.
2. I try to separate different functions into different method and break them to multiple functions. About renaming methods and function I will revise them.
3. I will remove commented codes.
4. I will remove extra comment with second language. Keep Farsi.
5. I will remove this function by I have issue about this function and raise the problem in this discussion (If possible).
6. I will correct the spell.
7. I will separate classes.
8. Certainly.
9. Certainly.

**Conclusion:**
1. Unfortunately I have never focused on Clean Architecture & Unit Test.
2. I am training by self study with reading books and watching video and thought that considering many factors may stop me from getting promotion in learning path.
3. I hope to implement all factors in an application to make a standard product.

thanks

**Reply by VahidFarahmandian** [reply]
@CesSolutions keep going and if you need help, you can count on me for sure.

**Comment by CesSolutions** [comment]
Dear @VahidFarahmandian
Should I ask question according to subjects in **ReadMe**? Can it be scattered and do not follow sequence of chapter?

**Reply by VahidFarahmandian** [reply]
Yeah.  Sure.  You can ask whatever question you have.  However, our suggestion is that the interns go step by step according to the weekly schedule.

**Comment by CesSolutions** [comment]
Hello Dear @VahidFarahmandian 

I am working on some projects to promote my programing skills but, I really want to follow **Clean Architecture**. As I know this architecture helps to manage and locate all objects of our solution in correct place and directory. In fact, this architecture defines solution break down. I searched a lot on the web and found some sample of that, but I really want to have the most comprehensive break down of solution to start to follow this architecture in my sample projects. In the next post I send my solution objects to get better suggestion from you. I don't know which directory and sub-directory must be defined in solution for projects, classes, ... .

thanks

**Reply by CesSolutions** [reply]
In all resources I found that **Clean Architecture** comes with **CQRS**. Is it necessary to follow **Clean Architecture** with **CQRS**? What about simple application? For Windows Form application and Asp.Net?

**Reply by VahidFarahmandian** [reply]
Every thing in software architecture is trade-off. So having CQRS+clean arch or whatever is a trade-off. You should check and verify what is beneficial for you application

**Comment by CesSolutions** [comment]
The following image is my solution and projects added without categorizing them into directories.
![Solution](https://github.com/VahidFarahmandian/spoota-internship-roadmap/assets/74654532/dc5421ab-f527-46ab-9079-622afd0da05b)

**Comment by CesSolutions** [comment]
Dear @VahidFarahmandian 

**Chapter:** Week01 > General > Working on Social Profiles

One of my biggest problem is English language and I have received negative feedback and warned me if I continue to ask question with grammatical mistake, my account will be blocked. But nowadays I am reader on StackOverflow. I never accepting Linked in with technical post because users specially beginners like me cannot gather technical information correctly and as Iranian I prefer to use [Barnamenevis.org](https://barnamenevis.org/) in which you can access users idea by category and all data is related to each other.

Thanks

**Comment by CesSolutions** [comment]
Dear @VahidFarahmandian 

**Chapter:** Week01 > General > Basic Concepts > What is version control

I use GitHub only to save my last state and be able to restore my project to previous point. I have never been engaged to something sophisticated to be enforced to use all features of GitHub. Would you please suggest me how to learn most practical and useful features of GitHub? Is there any simple resources to learn them by example?

Thanks

**Reply by VahidFarahmandian** [reply]
You need to read more about version control systems such as `git` and their differences compared to version control service providers such as GitHub

**Comment by CesSolutions** [comment]
Dear @VahidFarahmandian 

**Chapter:** Week01 > General > Basic Concepts > What is .Net

I am somehow familiar with .Net and learning about that continuously. My question is that, which third part application or technology I must learn beside .Net? Can I rely on .Net always? Would you please name some necessary third part application / technology which I should follow beside .Net? 

Thanks

**Comment by CesSolutions** [comment]
Dear @VahidFarahmandian 

Week 01 > General > Other Concepts > What is Software Architecture

**I have some questions about Value Objects in DDD:**
1. Each property in Entity/AggregateRoot has one validation check at least. Like **Description** field with **MaxLength**.
2. So, is it necessary that all properties in Entity/AggregateRoot shall be defined as ValueObjetcs instead of primitive types?
3. As I found, domain must not be dependent on outer world. So, where do I implement validation check without installing Nuget package? If I create value objects for all properties in Entity/AggregateRoot, So, validation is done inside ValueObject class and do not need to use extra validation check. Is it correct?

**Final Question:**
Is it possible to create Entity without ValueObjects then during development replace ValueObjects with primitive types.

Thanks

**Reply by VahidFarahmandian** [reply]
You should pay attention to the validation types. You need to perform Business related validations in your domain. DB related validations should be performed in DB section (such as in EF mapping configurations)

While using DDD it is highly required that you pay attention to Business logics and separate them from DB or other type logics.

**Comment by CesSolutions** [comment]
Dear @VahidFarahmandian 

Whould you please direct me about final question in my last comment? Thanks

**Reply by VahidFarahmandian** [reply]
There is nothing impossible. The point is that if a type in the form of business logic has the ValueObject charactristic, ValueObject must also be defined, otherwise you can put the logic related to the creation and parsing of ValueOvject inside the Entity.  For example, consider the GetAddress method, when called, puts the data related to the address together.  I emphasize again that in this example, if the business logic allows, Address should be the ValueObject. Note that this type of domain structure does not necessarily have anything to do with the database structure.

P.s: while using DDD, you are going to design the system based on the business and its literature. You should not design/write anything which is not in the bussiness literature.
